sympy/solvers/solveset.py

```diff
@@ -635,11 +634,12 @@ def _solveset(f, symbol, domain, _check=False):
             result = FiniteSet(*[s for s in result
                       if isinstance(s, RootOf)
                       or domain_check(fx, symbol, s)])
-
+    if _was_domain:
+        return Intersection(result, domain)
     return result
```

Comment:
Quoting from my previous comment :

> And [here](https://github.com/nishnik/sympy/blob/solveset_bug_fix/sympy/solvers/tests/test_solveset.py#l743), after #10448, 
```
assert solveset(sin(y + a) - sin(y), a, domain=S.Reals) == \
    Union(imageset(Lambda(n, 2*n*pi), S.Integers),
          imageset(Lambda(n, -I*(I*(2*n*pi + arg(-exp(-2*I*y)))
                                 + 2*im(y))), S.Integers))
```
This test case was passing, which I think is a wrong answer. This PR fixes it to:
```
assert solveset(sin(y + a) - sin(y), a, domain=S.Reals) == \
        imageset(Lambda(n, 2*n*pi), S.Integers)
```

So, if we just do :
```
if domain is S.Complexes or domain is S.Reals:
    return result
return Intersection(result, domain)
```
Again that test case will pass. It is because `solve_real_trig(f, symbol)` doesn't have a domain intersection thing.
I propose that in a long term, we can make every `solve_*` function do intersection independently. PS: `*` is used as a wild-card here.